ledger/onlineaccountscache.go

```diff
@@ -0,0 +1,111 @@
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package ledger
+
+import (
+	"github.com/algorand/go-algorand/data/basics"
+)
+
+const onlineAccountsCacheMaxSize = 1000
+
+type onlineAccountsCache struct {
+	accounts map[basics.Address]*persistedOnlineAccountDataList
+}
+
+// init initializes the onlineAccountsCache for use.
+// thread locking semantics : write lock
+func (o *onlineAccountsCache) init() {
+	o.accounts = make(map[basics.Address]*persistedOnlineAccountDataList)
+}
+
+// read the persistedAccountData object that the cache has for the given address.
+// thread locking semantics : read lock
+func (o *onlineAccountsCache) read(addr basics.Address, rnd basics.Round) (data persistedOnlineAccountData, has bool) {
+	if list := o.accounts[addr]; list != nil {
+		node := list.back()
+		if node.Value.round > rnd {
+			return persistedOnlineAccountData{}, false
+		}
+		for node.prev != &list.root {
+			node = node.prev
+			// only need one entry that is targetRound or older
+			if node.Value.round > rnd {
+				return *node.next.Value, true
+			}
+		}
+		return *node.Value, true
+	}
+	return persistedOnlineAccountData{}, false
+}
+
+// write a single persistedAccountData to the cache
+// thread locking semantics : write lock
+func (o *onlineAccountsCache) writeFront(acctData persistedOnlineAccountData) {
+	if _, ok := o.accounts[acctData.addr]; !ok {
+		if len(o.accounts) >= onlineAccountsCacheMaxSize {
+			return
+		}
+		o.accounts[acctData.addr] = newPersistedOnlineAccountList()
+	}
+	list := o.accounts[acctData.addr]
+	if list.root.next != &list.root && acctData.round <= list.root.next.Value.round {
+		return
+	}
+	o.accounts[acctData.addr].pushFront(&acctData)
+}
+
+// write a single persistedAccountData to the cache
+// thread locking semantics : write lock
+func (o *onlineAccountsCache) writeBack(acctData persistedOnlineAccountData) {
+	if _, ok := o.accounts[acctData.addr]; !ok {
+		if len(o.accounts) >= onlineAccountsCacheMaxSize {
+			return
+		}
+		o.accounts[acctData.addr] = newPersistedOnlineAccountList()
+	}
+	list := o.accounts[acctData.addr]
+	if list.root.prev != &list.root && acctData.round >= list.root.prev.Value.round {
+		return
+	}
+	o.accounts[acctData.addr].pushBack(&acctData)
+}
+
+// prune trims the onlineaccountscache by only keeping entries that would give account state
+// of rounds past targetRound
+// thread locking semantics : write lock
+func (o *onlineAccountsCache) prune(targetRound basics.Round) {
+	for addr, list := range o.accounts {
+		node := list.back()
+		for node.prev != &list.root {
+			node = node.prev
+			// only need one entry that is targetRound or older
+			// discard all older additional entries older than targetRound
+			if node.Value.round <= targetRound {
+				list.remove(node.next)
+			} else {
+				break
+			}
+		}
+		// only one item left in cache
+		if node.prev == &list.root && node.next == &list.root {
+			if node.Value.accountData.MsgIsZero() {
```

Comment:
I think VotingData should be checked. Plus need a test for this - checking offline accounts get pruned